File: backend/internal/delivery/performance.go
package delivery

import (
	"net/http"
	"strconv"

	"github.com/labstack/echo/v4"
	"github.com/typefunco/dealer_dev_platform/internal/utils"
)

// PerformanceDealerResponse представляет дилера с данными Performance для API.
type PerformanceDealerResponse struct {
	ID                  string  `json:"id"`
	Name                string  `json:"name"`
	City                string  `json:"city"`
	SrRub               string  `json:"srRub"`               // Sales Revenue
	SalesProfit         float64 `json:"salesProfit"`         // Sales Profit %
	SalesMargin         float64 `json:"salesMargin"`         // Sales Margin %
	AutoSalesRevenue    string  `json:"autoSalesRevenue"`    // After Sales Revenue
	Rap                 string  `json:"rap"`                 // RAP (можно вычислить позже)
	AutoSalesProfitsRap string  `json:"autoSalesProfitsRap"` // After Sales Profit
	AutoSalesMargin     float64 `json:"autoSalesMargin"`     // After Sales Margin %
	MarketingInvestment float64 `json:"marketingInvestment"` // Marketing Investment (M Rub)
	Ranking             int     `json:"ranking"`             // Foton Ranking
	AutoSalesDecision   string  `json:"autoSalesDecision"`   // Performance Decision
}

// GetPerformanceData возвращает данные производительности с поддержкой фильтров.
// @Summary Get Performance data
// @Description Получение данных производительности с фильтрацией по региону, году, кварталу и дилерам
// @Tags performance
// @Accept json
// @Produce json
// @Param region query string false "Region filter"
// @Param quarter query string false "Quarter filter (Q1, Q2, Q3, Q4)"
// @Param year query int false "Year filter"
// @Param dealer_ids query string false "Comma-separated dealer IDs"
// @Param limit query int false "Limit for pagination"
// @Param offset query int false "Offset for pagination"
// @Param sort_by query string false "Sort field"
// @Param sort_order query string false "Sort order (asc, desc)"
// @Success 200 {array} PerformanceDealerResponse
// @Failure 400 {object} ErrorResponse
// @Failure 500 {object} ErrorResponse
// @Router /api/performance [get]
func (s *Server) GetPerformanceData(c echo.Context) error {
	// Парсим параметры фильтрации с помощью универсальной функции
	filters := utils.ParseFilterParamsFromContext(c)

	// Устанавливаем значения по умолчанию для обратной совместимости
	defaults := map[string]interface{}{
		"region":  "all-russia",
		"quarter": "Q1",
		"year":    2024,
	}
	utils.SetDefaultFilters(filters, defaults)

	// Получение данных из сервиса с фильтрами
	perfList, err := s.perfService.GetPerformanceWithFilters(c.Request().Context(), filters)
	if err != nil {
		s.logger.Error("GetPerformanceData: failed to get performance data",
			"filters", filters,
			"error", err,
		)
		return c.JSON(http.StatusInternalServerError, ErrorResponse{
			Error: "Failed to get performance data",
		})
	}

	// Преобразование в API response
	response := make([]PerformanceDealerResponse, 0, len(perfList))
	for _, perf := range perfList {
		// RAP можно вычислить на основе различных метрик
		rap := calculateRAP(int16(perf.FotonRank))

		response = append(response, PerformanceDealerResponse{
			ID:                  strconv.FormatInt(int64(perf.DealerID), 10),
			Name:                perf.DealerNameRu,
			City:                perf.City,
			SrRub:               formatMoney(int64(perf.SalesRevenueRub)),
			SalesProfit:         perf.SalesProfitRub,
			SalesMargin:         perf.SalesMarginPercent,
			AutoSalesRevenue:    formatMoney(int64(perf.AfterSalesRevenueRub)),
			Rap:                 rap,
			AutoSalesProfitsRap: formatMoney(int64(perf.AfterSalesProfitRub)),
			AutoSalesMargin:     perf.AfterSalesMarginPercent,
			MarketingInvestment: perf.MarketingInvestment,
			Ranking:             perf.FotonRank,
			AutoSalesDecision:   perf.PerformanceDecision,
		})
	}

	return c.JSON(http.StatusOK, response)
}

// calculateRAP вычисляет RAP на основе рейтинга.
// Это упрощенная логика - в реальности может быть более сложная.
func calculateRAP(rank int16) string {
	switch {
	case rank >= 1 && rank <= 3:
		return "Gold"
	case rank >= 4 && rank <= 6:
		return "Silver"
	case rank >= 7 && rank <= 10:
		return "Bronze"
	default:
		return "None"
	}
}

File: frontend/src/api/performance.ts
// API клиент для работы с данными производительности
// Интеграция с backend Performance API

const API_BASE_URL = import.meta.env.DEV ? '/api' : 'http://localhost:8080/api'

export interface PerformanceDealer {
  id: string;
  name: string;
  city: string;
  srRub: string;
  salesProfit: number;
  salesMargin: number;
  autoSalesRevenue: string;
  rap: string;
  autoSalesProfitsRap: string;
  autoSalesMargin: number;
  marketingInvestment: number;
  ranking: number;
  autoSalesDecision: 'Needs Development' | 'Planned Result' | 'Find New Candidate' | 'Close Down';
}

export interface PerformanceResponse {
  dealers: PerformanceDealer[];
  pagination: {
    page: number;
    limit: number;
    total: number;
    totalPages: number;
  };
}

export interface PerformanceFilters {
  search?: string;
  region?: string;
  decision?: string;
  page?: number;
  limit?: number;
}

/**
 * Получить данные производительности дилеров
 */
export async function getPerformanceData(filters?: PerformanceFilters): Promise<PerformanceResponse> {
  const params = new URLSearchParams();
  
  if (filters?.search) params.append('search', filters.search);
  if (filters?.region) params.append('region', filters.region);
  if (filters?.decision) params.append('decision', filters.decision);
  if (filters?.page) params.append('page', filters.page.toString());
  if (filters?.limit) params.append('limit', filters.limit.toString());
  
  // Добавляем обязательные параметры quarter и year
  params.append('quarter', 'Q1');
  params.append('year', '2024');

  const response = await fetch(`${API_BASE_URL}/performance?${params.toString()}`, {
    credentials: 'include', // Включаем cookies для аутентификации
  });
  
  if (!response.ok) {
    throw new Error(`Failed to fetch performance data: ${response.statusText}`);
  }
  
  const dealers = await response.json();
  
  // Преобразуем ответ бэкенда в формат, ожидаемый фронтендом
  return {
    dealers: Array.isArray(dealers) ? dealers : [],
    pagination: {
      page: filters?.page || 1,
      limit: filters?.limit || 10,
      total: Array.isArray(dealers) ? dealers.length : 0,
      totalPages: Math.ceil((Array.isArray(dealers) ? dealers.length : 0) / (filters?.limit || 10))
    }
  };
}

// Хелперы

/**
 * Список доступных решений по производительности
 */
export const PERFORMANCE_DECISIONS = [
  'Planned Result',
  'Needs Development',
  'Find New Candidate',
  'Close Down'
] as const;

export type PerformanceDecision = typeof PERFORMANCE_DECISIONS[number];
